### PR TITLE
fix: kubectl plugin should use dynamic client

### LINKED ISSUE #808: kubectl plugin create should create objects using dynamic client
## Summary 

Currently the plugin is tightly coupled with a specific rollout version because it unmarshals files into the typed objects (which might be out-of-date from the version running in the cluster.

For example, if the controller is running a newer version of rollouts, and the plugin is outdated, during a create the plugin will drop all unknown fields which are not known by the plugin.

We should instead use the dynamic client and deal with unstructured objects. This way, an older plugin will still be forwards compatible with newer rollout versions

## Diagnostics

What version of Argo Rollouts are you running?

v0.9.x

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by khhirani**
Closes #808 

Signed-off-by: Kareena Hirani kareena_hirani@intuit.com

### CODE REVIEW

File: pkg/kubectl-argo-rollouts/cmd/create/create.go [hunk 1]
```diff
@@ -290,11 +290,17 @@ func NewCmdCreateAnalysisRun(o *options.ArgoRolloutsOptions) *cobra.Command {
...
 				return err
 			}
-			fmt.Fprintf(createOptions.Out, "analysisrun.argoproj.io/%s created\n", created.Name)
+			var un unstructured.Unstructured
+			unmarshal(runBytes, un)
+			_, err = createOptions.DynamicClient.Resource(v1alpha1.AnalysisRunGVR).Namespace(ns).Create(ctx, &un, metav1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(createOptions.Out, "analysisrun.argoproj.io/%s created\n", run.Name)
```
Comment (khhirani) [inline pkg/kubectl-argo-rollouts/cmd/create/create.go:303]: @jessesuen Did I convert the AnalysisRun to an unstructured object correctly?

**Review by jessesuen — CHANGES_REQUESTED** [review]
`analysisutil.NewAnalysisRunFromClusterTemplate` and `analysisutil.NewAnalysisRunFromTemplate` are two more instances where we will drop fields because we are getting a version of AnalysisRun which is tied to the client. We'll need a version of that which deals with, and returns an unstructured object, that the CLI will use.

**Comment by khhirani** [comment]
@jessesuen Would I convert the AnalysisTemplate/ClusterAnalysisTemplate to an Unstructured object, make changes to it, and then create it?

File: utils/analysis/helpers.go [hunk 2]
```diff
@@ -328,6 +329,41 @@ func flattenMetrics(templates []*v1alpha1.AnalysisTemplate, clusterTemplates []*
...
+
+	// Change kind to AnalysisRun
+	unstructured.SetNestedField(obj.Object, "AnalysisRun", "kind")
+
+	unstructured.SetNestedField(obj.Object, name, "metadata", "name")
+	unstructured.SetNestedField(obj.Object, generateName, "metadata", "generateName")
+	unstructured.SetNestedField(obj.Object, namespace, "metadata", "namespace")
+
+	// Set args
+	unstructured.SetNestedField(obj.Object, newArgs, "spec", "args")
```
Comment (khhirani) [inline utils/analysis/helpers.go:362]: @jessesuen This line is causing an error: ```panic: cannot deep copy *[]v1alpha1.Argument```

Comment (jessesuen) [reply utils/analysis/helpers.go:362]: yea you have to convert []v1alpha1.Argument to map[string]interface

**Review by jessesuen — APPROVED** [review]

**Pull request merged**